cpp/daal/src/algorithms/dtrees/forest/df_train_dense_default_impl.i

```diff
@@ -633,6 +748,188 @@ typename DataHelper::NodeType::Base * TrainBatchTaskBase<algorithmFPType, DataHe
     return makeLeaf(_aSample.get() + iStart, n, curImpurity, nClasses);
 }
 
+template <typename algorithmFPType, typename DataHelper, CpuType cpu>
+typename DataHelper::NodeType::Base * TrainBatchTaskBase<algorithmFPType, DataHelper, cpu>::buildBestFirst(
+    services::Status & s, size_t iStart, size_t n, size_t level, typename DataHelper::ImpurityData & curImpurity, bool & bUnorderedFeaturesUsed,
+    size_t nClasses)
+{
+    struct WorkItem
+    {
+        bool isLeaf;
+        bool featureUnordered;
+        size_t start;
+        size_t n;
+        size_t nLeft;
+        size_t level;
+        double improvement;
+        typename DataHelper::ImpurityData impurityLeft;
+        typename DataHelper::ImpurityData impurityRight;
+        typename DataHelper::NodeType::Split * node;
+
+        WorkItem() : isLeaf(true), featureUnordered(false), start(0), n(0), nLeft(0), level(0), improvement(0.0), node(nullptr) {}
+
+        WorkItem(bool featureUnordered, size_t start, size_t n, size_t level)
+            : isLeaf(true), featureUnordered(featureUnordered), start(start), n(n), nLeft(0), level(level), improvement(0.0), node(nullptr)
+        {}
+
+        void copy(WorkItem & src)
+        {
+            if (src.isLeaf)
+            {
+                improvement = 0.0;
+                isLeaf      = true;
+                return;
+            }
+
+            isLeaf           = src.isLeaf;
+            featureUnordered = src.featureUnordered;
+            start            = src.start;
+            n                = src.n;
+            nLeft            = src.nLeft;
+            level            = src.level;
+            improvement      = src.improvement;
+            impurityLeft     = src.impurityLeft;
+            impurityRight    = src.impurityRight;
+            node             = src.node;
+        }
+    };
+
+    if (_hostApp.isCancelled(s, n)) return nullptr;
+    BinaryHeap<WorkItem> binaryHeap(s);
+    if (!s.ok()) return nullptr;
+    size_t remainingSplitNodes = _maxLeafNodes - 1;
```

Comment:
`_maxLeafNodes` is zero as default case. we have `size_t` overflow. is this expected?

Reply:
_maxLeafNodes couldn't be zero here because it was checked above, line 632